### DISCUSSION TITLE
How to use Modelable /   Reactive attributes in my component ?

### DISCUSSION THREAD

**Discussion by sergeynilov** [livewire/livewire#8914] [1. Help] [answered]
A pair of months ago in laravel 11 / livewire 3 app I made votes system in component
UserReaction, where there were several buttons with current votes and where logged user could leave his vote,
where in  reactionActions I kept array with data on any post:

```
<?php

namespace App\Livewire;

use Livewire\Attributes\Reactive;

/*
 * component to show related UserReaction block of model(news, )
called as :
     @livewire('user-reaction', ['model'=>$news, 'modelType'=>\App\Models\News::class, 'showCommentsInfo' => true, 'showBottomBorder'=>false], key($news->id))
*
*/

class UserReaction extends Component
{
    use AppCommonTrait;

    /* @param  int - Model Id */
    public ?Model $model;

    /* @param  string - Model type with full path  like "\App\Models\News" */
    public string $modelType;

    # retrieved values from db - actions count for model
    #[Reactive]
    public array $reactionActions = [];

    public int $pollInSeconds = 0;

    /* @param  Model  $model  - model of reactions */
    public function mount(Model $model, string $modelType, b) {
        $this->model = $model;
        $this->modelType = $modelType;
        $this->retrieveDbData();
    }

    public function render(): View
    {
        \Log::info(varDump(-1, ' -1 UserReaction render::'));
        if($this->pollInSeconds > 0) { // need to refresh reactions/comments/complains data
            $this->retrievesReactionsCount();
        }
        return view('livewire.common.user-reaction', ['reactionActions' => $this->reactionActions]);
    }


    /*
    retrieve reactions count from db
    *
    * @return void
    */
    protected function retrievesReactionsCount()
    {
        $reactionTb = (new Reaction)->getTable();
        $totalUserReactions = Reaction // retrieve Reactions for current model
        ::getByModelType($this->modelType)
            ->getByModelId($this->model->id)
            ->groupBy('action')
            ->select($reactionTb.'.action',
                DB::raw('count('.$reactionTb.'.id) as reaction_count'))
            ->get()->toArray();

        $userReactions = [];
        if (auth()->user()) { // Votes by logged user must be defined in style(by ->voted property)
            $userReactions = Reaction::getByModelType($this->modelType)->getByModelId($this->model->id)->getByUserId(auth()->user()->id)->get()->pluck('action')->toArray();
        }
        // List of reaction type in the system
        $reactionActions = ReactionActionEnum::getSelectionItems();
        $this->reactionActions = [];
        foreach ($reactionActions as $key => $label) {  // collect all data in 1 array
            $voted = in_array($key, $userReactions);
            $reactionCount = 0;
            foreach ($totalUserReactions as $totalUserReaction) {
                if ($totalUserReaction['action'] === $key) {
                    $reactionCount = $totalUserReaction['reaction_count'];
                }
            }
            $this->reactionActions[$key] = [
                'key' => $key,
                'label' => Str::lower($label),
                'voted' => $voted, // if logged user has voted this action
                'reactionCount' => $reactionCount, // all users has voted this action
            ];
        }
    }

    /**
     * Method to add +1 action by user
     *
     * @param  int  $action  - which action was added
     *
     * @return void
     */
    public function voteReaction(int $action)
    {
        if ( ! LoggedUserFacade::checkUserLogged()) {
            self::dispatchAlert(DispatchAlertTypeEnum::WARNING, 'You need to login firstly to react !');

            return;
        }
        $reaction = Reaction::getByModelType($this->modelType)->getByModelId($this->model->id)->getByAction($action)->getByUserId(auth()->user()->id)->first();
        if (empty($reaction)) { // not to vote for the second time
            Reaction::create([
                'model_id' => $this->model->id,
                'model_type' => $this->modelType,
                'user_id' => auth()->user()->id,
                'action' => $action
            ]);
            self::dispatchAlert(DispatchAlertTypeEnum::SUCCESS, 'Your reaction was added !');
            $this->reactionActions[$action]['voted'] = true;
        }
        $this->retrievesReactionsCount();
    }

```
    When logged user clicked on some button voteReaction method is clicked and method retrievesReactionsCount
was called to retrieve from db data for reaction buttons, without reloading all page.

It worked ok untill last time, when I got error :

`    Cannot mutate reactive prop [reactionActions] in component: [user-reaction]
`
Searching in net I found a decision to to change attribute :

```
    #[Modelable]
//    #[Reactive]
    public array $reactionActions = [];

```
Now it works ok. Why such changes. Not sure is it after some upgrade ?

Reading https://livewire.laravel.com/docs/nesting#reactive-props I am confdused : 
I have no parent components, why such errors ?

**Comment by EdgarsJoja** [comment, marked as answer]
I'm not sure you need any modifier for `$reactionActions` property at all.
As you said, you have no parent component, so this does not need to be a reactive prop - it shouldn't affect anything. Reactive props are useful when there is component nesting and you want the prop to automatically update when parent changes it. This is not the case here, if I understand correctly.
And modelable is used when you essentially want to expose custom "model" state, for some child component - again, since you wrote that there is no parent component, then this is unnecessary and won't do anything.
Adding custom modifiers add some restrictions (i.e. your error), so they should be avoided unless it's actually needed.

tldr; I think you don't need any modifiers here:
```
public array $reactionActions = [];
```
and that's it.
